### PR TITLE
Allow bucketing by month, year, century in time_bucket and time_bucket_gapfill

### LINKED ISSUE #4005: Implement time_bucket_gapfill_ng()
### What type of enhancement is this?

User experience

### What subsystems and features will be improved?

Gapfill

### What does the enhancement do?

Right now to do time_buckets by non fixed times e.g yearly, monthly and quarterly we need to use timescaledb_experimental.time_bucket_ng. Which seems to work. But I can't seem to find any way to gapfill results like time_bucket_gapfill. I would expect a similar function: timescaledb_experimental.time_bucket_ng_gapfill

SME: @svenklemm 

### Implementation challenges

what time_bucket_gapfill is to time_bucket should timescaledb_experimental.time_bucket_ng_gapfill be to timescaledb_experimental.time_bucket_ng

- [ ] Tests: https://github.com/timescale/product/issues/147

---

### References
- https://github.com/timescale/timescaledb/issues/1324
- https://github.com/timescale/timescaledb/issues/1435
- https://github.com/timescale/timescaledb/issues/3212

### PR DESCRIPTION
**Pull request by svenklemm**
This patch allows bucketing by month for time_bucket with date,
timestamp or timestamptz. When bucketing by month the interval
must only contain month components.

Fixes #4005

Disable-check: commit-count

### CODE REVIEW

**Review by akuzm — APPROVED** [review]
Looks like the gapfill tests are missing.

File: src/time_bucket.c [hunk 1]
```diff
@@ -134,6 +134,43 @@ ts_int64_bucket(PG_FUNCTION_ARGS)
...
+
+	if (interval->day || interval->time)
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("month intervals cannot have day or time component")));
+	}
+}
+
+static DateADT
```
Comment (gayyappan) [inline src/time_bucket.c:153]: Could you add some highlevel comments about this function?

File: tsl/src/nodes/gapfill/exec.c [hunk 2]
```diff
@@ -160,7 +160,13 @@ gapfill_period_get_internal(Oid timetype, Oid argtype, Datum arg)
 		case TIMESTAMPOID:
 		case TIMESTAMPTZOID:
 			Assert(INTERVALOID == argtype);
-			return interval_to_usec(DatumGetIntervalP(arg));
+			Interval *interval_arg = DatumGetIntervalP(arg);
+			if (interval_arg->month)
+			{
```
Comment (gayyappan) [inline tsl/src/nodes/gapfill/exec.c:165]: need a test for gapfill with months?

File: tsl/src/nodes/gapfill/exec.h [hunk 3]
```diff
@@ -96,6 +96,7 @@ typedef struct GapFillState
 	int64 gapfill_start;
 	int64 gapfill_end;
 	int64 gapfill_period;
+	Interval *gapfill_interval;
```
Comment (gayyappan) [inline tsl/src/nodes/gapfill/exec.h:99]: would be good to add a comment about how gapfill_period and gapfill_interval are used.

**Review by gayyappan — APPROVED** [review]
minor comments.

**Review by konskov — COMMENTED** [review]
I think the error message in `get_interval_period_timestamp_units` ("interval defined in terms of month, year, century etc. not supported") can now be updated by removing the "month" part

**Pull request merged**